### ISSUE TITLE
TextSpan tree accessibility improvements

### ISSUE THREAD

**Issue by mehmetf** [flutter/flutter#36306]
Internal bug: b/137602251

We would like to be able to show a message such as:

"By continuing, you agree to our _Terms of Service_ and _Privacy Policy_."

We would like behavior to existing `NSAttributedString` on iOS:

a) The entire text can be selected at once, and will read out "By continuing, you agree to our Terms of Service (Link) and Privacy Policy (link)

b) Swiping right in link mode* will then take you to "Terms of Service". Swiping right again should take you to "Privacy Policy".

Right now, it seems that RichText currently splits up the semantics into 4 parts by default, and will only read out "By continuing, you agree to our ". Then you have to swipe right to read "Terms of Service", then right again for "and", and then a final time for "Privacy Policy"

* see https://a11y-guidelines.orange.com/mobile_EN/voiceover.html#OpenLink

**Comment by QuattroNinos** [comment]
Internally tracked in b/140825718

**Comment by tvolkert** [comment]
Customer Money (g3) needs this fixed by mid-Q4.  Removing the milestone for now since it's not yet assigned.

**Comment by goderbauer** [comment]
What we currently do matches how text and links are represented in Safari on iOS and Chrome on Android.

Steps:

1) Check that native actually offers this other behavior (on both platforms?)
2) We could provide a flag on Text to switch between the two behaviors
3) We may have to add a SemanticsFlag for Links to match https://developer.apple.com/documentation/uikit/uiaccessibility/uiaccessibilitytraits/1620178-link if that's what iOS uses (check if there's an equivalent for Android).

**Comment by jonahwilliams** [comment]
There will likely be similar requirements for link accessibility on the web.

**Comment by goderbauer** [comment]
@jonahwilliams What do you mean? Isn't what we currently do basically what the web does?

**Comment by jonahwilliams** [comment]
What I mean is that we have no concept of a `link` in our semantics.

**Comment by goderbauer** [comment]
Oh, yes. Makes sense.

**Comment by Piinks** [comment]
This is a runnable sample I'm working with to test behavior based on the description above. If this reproduction is off-base let me know. :)

```dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:url_launcher/url_launcher.dart' as launcher;

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) => MaterialApp(home: MyHomePage(title: 'Demo'));
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final String flutter = 'https://flutter.dev/';
  final String pub = 'https://pub.dev/';
  
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(widget.title)),
      body: Center(
        child: RichText(
          text: TextSpan(
            style: TextStyle(color: Colors.black87),
            children: <TextSpan>[
              TextSpan(text: 'By continuing, you agree to our '),
              TextSpan(
                text: 'Terms of Service',
                style: TextStyle(color: Colors.blue),
                recognizer: TapGestureRecognizer()
                  ..onTap = () => launcher.launch(flutter),
              ),
              TextSpan(text: ' and '),
              TextSpan(
                text: 'Privacy Policy',
                style: TextStyle(color: Colors.blue),
                recognizer: TapGestureRecognizer()
                  ..onTap = () => launcher.launch(pub),
              ),
              TextSpan(text: '.'),
            ],
          ),
        )
      ),
    );
  }
}

```

**Comment by Piinks** [comment]
cc/ @GaryQian who recently refactored `TextSpan`.

**Comment by goderbauer** [comment]
Customer reported this fixed.
